src/lib/apply-shim.html

```diff
@@ -0,0 +1,179 @@
+<!--
+@license
+Copyright (c) 2014 The Polymer Project Authors. All rights reserved.
+This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+-->
+<link rel="import" href="style-util.html">
+<script>
+Polymer.ApplyShim = (function(){
+  'use strict';
+
+  var styleUtil = Polymer.StyleUtil;
+
+  var MIXIN_MATCH = styleUtil.rx.MIXIN_MATCH;
+  var VAR_ASSIGN = styleUtil.rx.VAR_ASSIGN;
+  var VAR_MATCH = styleUtil.rx.VAR_MATCH;
+  var APPLY_NAME_CLEAN = /;\s*/m;
+
+  // map of mixin to property names
+  // --foo: {border: 2px} -> (--foo, ['border'])
+  var mixinMap = {};
+
+  function mapSet(name, prop) {
+    name = name.trim();
+    mixinMap[name] = prop;
+  }
+
+  function mapGet(name) {
+    name = name.trim();
+    return mixinMap[name];
+  }
+
+  // "parse" a mixin definition into a map of properties and values
+  // cssTextToMap('border: 2px solid black') -> ('border', '2px solid black')
+  function cssTextToMap(text) {
+    var props = text.split(';');
+    var out = {};
+    for (var i = 0, p, sp; i < props.length; i++) {
+      p = props[i];
+      if (p) {
+        sp = p.split(':');
+        // ignore lines that aren't definitions like @media
+        if (sp.length > 1) {
+          // some properties may have ':' in the value, like data urls
+          out[sp[0].trim()] = sp.slice(1).join(':');
+        }
+      }
+    }
+    return out;
+  }
+
+  function produceCssProperties(matchText, propertyName, valueProperty, valueMixin) {
+    // handle case where property value is a mixin
+    if (valueProperty) {
+      VAR_MATCH.lastIndex = 0;
+      var m = VAR_MATCH.exec(valueProperty);
+      if (m) {
+        var value = m[2];
+        if (mapGet(value)){
+          valueMixin = '@apply ' + value + ';';
+        }
+      }
+    }
+    if (!valueMixin) {
+      return matchText;
+    }
+    var mixinAsProperties = consumeCssProperties(valueMixin);
+    var prefix = matchText.slice(0, matchText.indexOf('--'));
+    var mixinValues = cssTextToMap(mixinAsProperties);
+    var oldProperties = mapGet(propertyName);
+    var combinedProps = mixinValues;
+    if (oldProperties) {
+      combinedProps = Polymer.Base.mixin(oldProperties, mixinValues);
+    } else {
```

Comment:
Since we are using `mixin`, the reference in the map is always the same and we don't need to `set` again.